### PR TITLE
[casr-cluster] Support deterministic clustering

### PR DESCRIPTION
**Pull request by hkctkuy**

### CODE REVIEW

File: casr/src/bin/casr-cluster.rs [hunk 1]
```diff
@@ -63,37 +63,28 @@ fn make_clusters(
...
         .build()
         .unwrap();
 
-    // Stacktraces from casreps
-    let traces: RwLock<Vec<Stacktrace>> = RwLock::new(Vec::new());
-    // Crashlines from casreps
-    let crashlines: RwLock<Vec<String>> = RwLock::new(Vec::new());
-    // Casreps with stacktraces, that we can parse
-    let filtered_casreps: RwLock<Vec<PathBuf>> = RwLock::new(Vec::new());
+    // Report info from casreps: (casrep, (trace, crashline))
```
Comment (SweetVishnya) [inline casr/src/bin/casr-cluster.rs:72]: Maybe, just a flat tuple with 3 elements

File: casr/src/bin/casr-cluster.rs [hunk 2]
```diff
@@ -63,37 +63,28 @@ fn make_clusters(
...
     custom_pool.install(|| {
         (0..len).into_par_iter().for_each(|i| {
             if let Ok(report) = util::report_from_file(casreps[i].as_path()) {
                 if let Ok(trace) = report.filtered_stacktrace() {
-                    traces.write().unwrap().push(trace);
-                    filtered_casreps.write().unwrap().push(casreps[i].clone());
+                    let mut crashline = String::new();
                     if dedup {
-                        crashlines.write().unwrap().push(report.crashline);
+                        crashline = report.crashline;
```
Comment (SweetVishnya) [inline casr/src/bin/casr-cluster.rs:82]: we may always save crash line, that's not much time and memory consuming

File: libcasr/src/stacktrace.rs [hunk 3]
```diff
@@ -239,6 +239,10 @@ pub fn cluster_stacktraces(stacktraces: &[Stacktrace]) -> Result<Vec<usize>> {
         counter += 1;
     }
 
+    // Sort clusters by keys
```
Comment (SweetVishnya) [inline libcasr/src/stacktrace.rs:242]: @Avgor46, please, check this part.

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: casr/src/bin/casr-cluster.rs (hunk 1 above)
Comment (hkctkuy) [reply casr/src/bin/casr-cluster.rs:72]: It's made according to `unzip()` documentation
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.unzip

Comment (hkctkuy) [reply casr/src/bin/casr-cluster.rs:72]: For triple tuple get tree vectors is not such simple

File: casr/src/bin/casr-cluster.rs (hunk 2 above)
Comment (hkctkuy) [reply casr/src/bin/casr-cluster.rs]: Done

File: casr/src/bin/casr-cluster.rs [hunk 4]
```diff
@@ -77,10 +77,7 @@ fn make_clusters(
         (0..len).into_par_iter().for_each(|i| {
             if let Ok(report) = util::report_from_file(casreps[i].as_path()) {
                 if let Ok(trace) = report.filtered_stacktrace() {
-                    let mut crashline = String::new();
-                    if dedup {
-                        crashline = report.crashline;
-                    }
+                    let crashline = report.crashline;
```
Comment (SweetVishnya) [inline casr/src/bin/casr-cluster.rs:80]: this var is used just once, you may emplace it

File: casr/src/bin/casr-cluster.rs (hunk 1 above)
Comment (SweetVishnya) [reply casr/src/bin/casr-cluster.rs:72]: ok

File: casr/src/bin/casr-cluster.rs (hunk 4 above)
Comment (hkctkuy) [reply casr/src/bin/casr-cluster.rs]: Done

File: libcasr/src/stacktrace.rs (hunk 3 above)
Comment (Avgor46) [reply libcasr/src/stacktrace.rs:242]: it should work correctly

**Review by SweetVishnya — APPROVED** [review]

File: casr/src/bin/casr-cluster.rs [hunk 5]
```diff
@@ -102,11 +89,21 @@ fn make_clusters(
             }
         })
     });
-    let stacktraces = traces.read().unwrap();
-    let crashlines = crashlines.read().unwrap();
-    let casreps = filtered_casreps.read().unwrap();
+    let casrep_info = casrep_info.get_mut().unwrap();
     let badreports = badreports.get_mut().unwrap();
 
+    // Sort by casrep path
```
[the remark this answers is not in the document]
Comment (hkctkuy) [reply casr/src/bin/casr-cluster.rs]: Done

**Review by Avgor46 — APPROVED** [review]

**Pull request merged**